src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs

```diff
@@ -195,66 +195,93 @@ public override void VisitParenthesizedLambdaExpression(ParenthesizedLambdaExpre
 
         public override void VisitLocalFunctionStatement(LocalFunctionStatementSyntax node)
         {
-            var body = (CSharpSyntaxNode)node.Body ?? node.ExpressionBody;
-            LocalFunctionSymbol match = null;
-            // Don't use LookupLocalFunction because it recurses up the tree, as it
-            // should be defined in the directly enclosing block (see note below)
+            LocalFunctionSymbol match = FindLocalFunction(node, _enclosing);
 
-            Binder possibleScopeBinder = _enclosing;
-            while (possibleScopeBinder != null && !possibleScopeBinder.IsLocalFunctionsScopeBinder)
-            {
-                possibleScopeBinder = possibleScopeBinder.Next;
-            }
+            bool oldSawYield = _sawYield;
+            _sawYield = false;
 
-            if (possibleScopeBinder != null)
+            BlockSyntax blockBody = node.Body;
+            ArrowExpressionClauseSyntax arrowBody = node.ExpressionBody;
+            var body = (CSharpSyntaxNode)blockBody ?? arrowBody;
+
+            InMethodBinder inMethodBinder = null;
+            if (body != null)
             {
-                foreach (var candidate in possibleScopeBinder.LocalFunctions)
+                VisitLocalFunctionBody(node, body, match, ref inMethodBinder);
```

Comment:
>ref inMethodBinder [](start = 58, length = 18)

This feels too complicated than it could be. We can calculate the binder right here. If we have a match, we create InMethodBinder, otherwise the binder is  _enclosing and we always pass the binder to the Visit method, for both bodies.

Reply:
Thanks. That's better indeed.

---
In reply to: [166453617](https://github.com/dotnet/roslyn/pull/24651#discussion_r166453617) [](ancestors = 166453617)